### ISSUE TITLE
Include numbering in PDF bookmark

### ISSUE THREAD

**Issue by laurmaedje** [typst/typst#2416]
### Description

Have the PDF outline bookmark contain the numbering.

### Use Case

Common in other software.

**Comment by laurmaedje** [comment]
See also: https://github.com/typst/typst/discussions/1163

**Comment by PgBiel** [comment]
Does PDF have a dedicated feature for bookmark numbering, or would it have to be part of the bookmark itself?

**Comment by laurmaedje** [comment]
I think it would need to be part of the bookmark / outline entry itself.

**Comment by caemor** [comment]
That would be really useful for us as well, we have the same heading names in the appendix to add more information to a few topics. Our normal headings are numbered normally (1,2,3...) and our appendix is using letters (A,B,...). But sadly this doesn't show up in the bookmarks, but it works in the outline.

**Comment by caemor** [comment]
And the corresponding output:

![grafik](https://github.com/typst/typst/assets/11088935/2e714382-bc0a-49d4-88b8-eebabd7af55f)

**Comment by SUPERCILEX** [comment]
Copied from #2306 

---

### Description

For example suppose you just want numbers in your heading: `align(center, pad(heading(numbering: "1", "")))`. Then the title of the heading is technically blank. When the PDF is exported, these blank titles are copied into the outline.

Looking at how textbooks do things, I believe the correct implementation would be to make the PDF heading match the rendered title (i.e. it would be the concatenation of the numbering and title). This improves readability of your typical titles as a corollary fixes the issue with blank PDF outlines because they will still have numbers.

**Comment by voidiz** [comment]
Hi! I'm new to the code base. I've been trying to look at this but the finalized number added as a prefix to the heading text doesn't seem to be a part of the `HeadingElem`, so I haven't found a way to retrieve it with `Introspector::query`. Instead, it seems to be dynamically created using `Counter::at` and `Counter::display`. However, these require an instance of `Engine` which afaik isn't available in the phase where `Document` is turned into a pdf.

Do you have any ideas on how the number should be retrieved? Should it be stored in the `HeadingElem` somehow? @laurmaedje

**Comment by laurmaedje** [comment]
@voidiz This is indeed tricky. I don't immediately see a clean way to solve this, I need to think more about it.

**Comment by unclebean** [comment]
Hello, I'm interesting to contribute to typst, since this issue is in the **Good first issues**, so I did some research, please have a look the screenshot, for my understanding is that we should add page number with the bookmark.

<img width="1071" alt="image" src="https://github.com/typst/typst/assets/8064750/12d41498-25c1-4f3b-8c3c-d7c9d4f7c5e0">

**Comment by YDX-2147483647** [comment]
> add page number with the bookmark

Hi! What we want is to include _section_ numbers, instead of _page_ numbers.

Example: Manual of the `amssymb` (LaTeX) package.

![](https://github.com/typst/typst/assets/73375426/9d51ace3-c9f6-4355-aff2-87b9ce394d9e)

**Comment by laurmaedje** [comment]
I've removed the good first issue label since it turns out to be a bit trickier to resolve the numbering. I'll try to look into what would be necessary to fix this (this time for real).

**Comment by laurmaedje** [comment]
The problem is that in the export stage we don't have access to the World/Engine etc. which is needed to resolve numbers + a numbering pattern or function into a concrete string/content.

I see two possible avenues to deal with this:
- Only support numbering patterns, not functions, as those can be resolved without an engine (the engine is only needed to call the numbering function)
- Resolve the numbering into an `#[internal]` field in the heading's `Synthesize` implementation and use that. It's a bit hacky, but I think it should work.

**Comment by unclebean** [comment]
Hi @laurmaedje sorry I spend some time to learn typst and try to resolve the issue so I didn't check your comments, my idea is just add numbers in typst-pdf/src/outline.rs **write_outline_item** function. I did basic testing seems is working fine. If you don't mind I can create PR for this issue.

<img width="1021" alt="image" src="https://github.com/typst/typst/assets/8064750/dda43f8f-92f6-4cae-9323-02fc6e2d3634">

**Comment by laurmaedje** [comment]
@unclebean how did you create the numbers? The big challenge is to have it properly interact with both `counter(heading)` updates and user-defined numberings.

**Comment by unclebean** [comment]
Hi @laurmaedje I have created one [PR](https://github.com/typst/typst/pull/3908) for it. 
My assumption is that in the same level user may define the header like this 

#set heading(numbering: "A.")
= Introduction
= Background
== History
== State of the Art

then in the **write_outline_item** function we can get **Numbering::Pattern** from **HeaderElem** since the iteration is following header sequence so we can create the numbers to call **Numbering::Pattern** apply function:

= Introduction             -> [1]
= Background             -> [2]
== History                   -> [2, 1]
== State of the Art     -> [2, 2]

after that we will get the number with bookmark like this:
<img width="828" alt="image" src="https://github.com/typst/typst/assets/8064750/4b4b3883-c629-4776-a47e-46a774f853d0">

**Comment by laurmaedje** [comment]
The problem is that this will ignore counter updates. Try putting `counter(heading).update(1)` somewhere in between the headings. The output will be different between PDF and `#outline()`.

**Comment by laurmaedje** [comment]
That probably works quite often, but is too brittle in my opinion. I think we should aim for a proper solution, which will likely involve a synthesized field, as said above. (I updated my comment to reflect that actually the counters are also a problem, not just non-pattern numberings.)

**Comment by unclebean** [comment]
I see, when I learn typst and try to find out the way to resolve the issue, I think may should have easy way to store number value rather than only have Pattern in Numbering. If my PR is not good to merge may I check with you any other issue I can pickup?

**Comment by goyalyashpal** [comment]
> See also: [#1163](https://github.com/typst/typst/discussions/1163)
> \- @ laurmaedje at https://github.com/typst/typst/issues/2416#issuecomment-1766114268

can u include the screenshot in the issue description itself please?

**Comment by YDX-2147483647** [comment]
> Resolve the displayed counter into an `#[internal]` field in the heading's `Synthesize` implementation and use that. It's a bit hacky, but I think it should work.
https://github.com/typst/typst/issues/2416#issuecomment-2042229942

If I or someone else manages to develop a PR implementing this solution, would it be merged? Is further discussion needed? 

In the meantime, here's a workaround (for typst v0.13.1) from https://forum.typst.app/t/how-to-display-chapter-numbers-in-pdf-bookmarks/4912.

**Update**:
- If you need to reference your heading, please use bluss' version in that post instead.
- If not all headings in your document is numbered (e.g. bibliography), please use SwampertX's version in that post instead.

```typ
// Remove original headings from bookmarks. We will add new ones later.
#set heading(bookmarked: false)

#show heading: it => if it.numbering == none {
  // This heading has been processed. Keep it untouched.
  it
} else {
  let (numbering, body, ..args) = it.fields()

  // Render the numbering manually
  let numbered-body = block({
    counter(heading).display(numbering)

    h(0.3em) // Space in the real document
    box(width: 0em, hide(" ")) // Space in the bookmark
    // You can remove the above spaces if you meet https://github.com/typst/typst/issues/5778.

    body
  })

  // Revert non-idempotent show rules.
  // https://github.com/typst/typst/blob/22a57fcf5c52cc0e961368a6d6fb7e9a82312b48/crates/typst-library/src/model/heading.rs#L309
  set text(size: 1em / 1.4) if it.depth == 1
  set text(size: 1em / 1.2) if it.depth == 2

  // Add our bookmarked heading
  heading(..args, bookmarked: true, numbering: none, numbered-body)
}


#set heading(numbering: "1.1")

= A

== B

= C

#counter(heading).update(0)
= D
```

<img width="363" height="375" alt="Image" src="https://github.com/user-attachments/assets/34305edd-c00c-4cd0-8e39-aa13ec7caa41" />

**Comment by laurmaedje** [comment]
> If I or someone else manages to develop a PR implementing this solution, would it be merged?

If it just works with a synthesized field, I'd say it has a good chance of being merged.

**Comment by alex-touza** [comment]
I think I'm experiencing this issue again in 0.15.1. It happens for numbering that comes from a counter or state other than `heading`'s own `numbering` field.

```typ
#counter("a")

= This heading is f1ne

= #context counter("a").display(). this one's not

#sys.version
```

In all PDF viewers I've tried, the bookmark for the second heading _doesn't_ include the `0`:

<img width="597" height="410" alt="Image" src="https://github.com/user-attachments/assets/39573ad5-1170-43a6-8176-a76f1e2478d8" />

It happens with both `state` and `counter`. 

I believe this issue should be reopened, as the rendered heading doesn't match the bookmark in this case.

**Comment by alex-touza** [comment]
Nevermind. The issue I'm talking about is actually https://github.com/typst/typst/issues/3424.

**Comment by YDX-2147483647** [comment]
Yes, and `#context [= #counter("a").display(). this]` is a workaround.

**Comment by alex-touza** [comment]
Yes, that is a workaround, the problem is if you have a contextual function, you have to inline the function. Not very bad, but a bit ugly indeed.

In my case I have a function that returns a localised message:

```typ
///
/// Gets a localised message from an ID.
/// - id (str):
/// - cap (bool): Whether to make the first letter uppercase.
/// -> content
#let m(id, cap) = context {
  let s = files.at(text.lang).at(id)
  return if cap {utils.capitalize(s)} else {s}
}
```

Imagine if I had a much longer function and I used it everywhere...
